packages/arb-avm-cpp/avm_values/src/buffer.cpp

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright 2019-2020, Offchain Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <avm_values/buffer.hpp>
+#include <avm_values/bigint.hpp>
+
+#include <ethash/keccak.hpp>
+
+uint256_t zero_hash(int sz) {
+    if (sz == 32) {
+        return hash(0);
+    }
+    auto h1 = zero_hash(sz/2);
+    return hash(h1, h1);
+}
+
+Packed normal(uint256_t hash, int sz) {
+    return Packed{hash, sz, 0};
+}
+
+Packed pack(const Packed& packed) {
+    return Packed{packed.hash, packed.size, packed.packed+1};
+}
+
+bool is_zero_hash(const Packed& packed) {
+    return packed.hash == hash(0);
+}
+
+uint256_t unpack(const Packed &packed) {
+    uint256_t res = packed.hash;
+    int sz = packed.size;
+    for (int i = 0; i < packed.packed; i++) {
+        res = hash(res, zero_hash(sz));
+        sz = sz*2;
+    }
+    return res;
+}
+
+Packed zero_packed(int sz) {
+    if (sz == 32) {
+        return normal(zero_hash(32), 32);
+    }
+    return pack(zero_packed(sz/2));
+}
+
+Packed hash_buf(uint8_t *buf, int offset, int sz) {
+    if (sz == 32) {
+        auto hash_val = ethash::keccak256(buf+offset, 32);
+        uint256_t res = intx::be::load<uint256_t>(hash_val);
+        return normal(res, 32);
+    }
+    // std::cerr << "hashing " << offset << " to " << (offset+sz) << std::endl;
+    auto h1 = hash_buf(buf, offset, sz/2);
+    auto h2 = hash_buf(buf, offset+sz/2, sz/2);
+    if (is_zero_hash(h2)) {
+        return pack(h1);
+    }
+    return normal(hash(unpack(h1), unpack(h2)), sz);
+}
+
+Packed hash_node(RawBuffer *buf, int offset, int len, int sz) {
```

Comment:
It looks like this function only ever reads from the first element of the buf array. Maybe the intention was to increment it as part of the recursion?